src/computational_graph/eval.jl

```diff
@@ -1,95 +1,25 @@
-@inline apply(o::Sum, diags::Vector{Diagram{W}}) where {W<:Number} = sum(d.weight for d in diags)
-@inline apply(o::Prod, diags::Vector{Diagram{W}}) where {W<:Number} = prod(d.weight for d in diags)
-@inline apply(o::Sum, diag::Diagram{W}) where {W<:Number} = diag.weight
-@inline apply(o::Prod, diag::Diagram{W}) where {W<:Number} = diag.weight
-
-@inline eval(d::DiagramId) = error("eval for $d has not yet implemented!")
-
-######################### evaluator for KT representation ######################### 
-function evalDiagNodeKT!(diag::Diagram, varK, varT, additional=nothing; eval=DiagTree.eval)
-    if length(diag.subdiagram) == 0
-        # if hasproperty(diag.id, :extK)
-        if (isnothing(varK) == false) && (isnothing(varT) == false)
-            K = varK * diag.id.extK
-            if isnothing(additional)
-                diag.weight = eval(diag.id, K, diag.id.extT, varT) * diag.factor
-            else
-                diag.weight = eval(diag.id, K, diag.id.extT, varT, additional) * diag.factor
-            end
-        elseif isnothing(varK)
-            if isnothing(additional)
-                diag.weight = eval(diag.id, diag.id.extT, varT) * diag.factor
-            else
-                diag.weight = eval(diag.id, diag.id.extT, varT, additional) * diag.factor
-            end
-        elseif isnothing(varT)
-            K = varK * diag.id.extK
-            if isnothing(additional)
-                diag.weight = eval(diag.id, K) * diag.factor
-            else
-                diag.weight = eval(diag.id, K, additional) * diag.factor
-            end
+#@inline add( diags::Vector{Graph{F,W}}) where {F<:Number,W<:Number} = sum(d.weight for d in diags)
+@inline apply(::Type{Sum}, diags::Vector{Graph{F,W}}, factors::Vector{F}) where {F<:Number,W<:Number} = sum(d.weight*d.factor*f for (d,f) in zip(diags,factors) )
+@inline apply(::Type{Prod}, diags::Vector{Graph{F,W}}, factors::Vector{F}) where {F<:Number,W<:Number} = prod(d.weight*d.factor*f for (d,f) in zip(diags,factors) )
+@inline apply(o::Sum, diag::Graph{F,W}) where {F<:Number,W<:Number} = diag.weight
+@inline apply(o::Prod, diag::Graph{F,W}) where {F<:Number,W<:Number} = diag.weight
+
+# #@inline eval(d::DiagramId) = error("eval for $d has not yet implemented!")
+# #
+function evalGraph!(g::Graph)
```

Comment:
rename evalGraph! -> eval!

Reply:
passing an additional vector "leaf" which contains the leaf weight.

eval! should have the same behavior as the compiled function